lib/rpmsg/rpmsg_virtio.c

```diff
@@ -333,6 +333,8 @@ static void *rpmsg_virtio_get_tx_payload_buffer(struct rpmsg_device *rdev,
 		metal_mutex_release(&rdev->lock);
 		if (rp_hdr || !tick_count)
 			break;
+		if (rpmsg_virtio_wait_tx_buffer(rvdev) >= 0)
```

Comment:
This looks wrong. If it fails, the program waits twice, once in rpmsg_virtio_wait_tx_buffer(), and then in metal_sleep_usec().

Reply:
If wait fails, that means hasn't wait, then we still need do metal_sleep_usec(), or will busyloop.

Reply:
Please could you explain why you don't use `rpmsg_trysend` API instead  that set `wait` to `false` and so `tick_count` to `0`?

Reply:
There is only one difference between a looping query mechanism and a waiting mechanism.
Of course, a waiting mechanism is more suitable for low power device.

Reply:
My point here is that I don't understand why you don't use "rpmsg_trysend" in an application loop instead.
you would generate your loop without waiting mechanism.

Reply:
But,  rpmsg_send/rpmsg_get_tx_payload_buffer is a public api, we can't limit how other team write the application.

Reply:
> If wait fails, that means hasn't wait, then we still need do metal_sleep_usec(), or will busyloop.

So a function with "wait" in its name doesn't wait sometimes? This needs to be documented in the Doxygen comments. There needs to be enough documentation that a new user of the library can understand why they might need to implement this operation, and how to implement it properly.

Reply:
> My point here is that I don't understand why you don't use "rpmsg_trysend" in an application loop instead.
> you would generate your loop without waiting mechanism.
I agree with @arnopo. I don't understand why this needs to be in the library, rather than in application code.

Reply:
> > If wait fails, that means hasn't wait, then we still need do metal_sleep_usec(), or will busyloop.
> 

Yes, wait fail mean it can't finish the work in some special case, so the common code need do the fallback action. In most case, wait success mean the buffer available and the framework can try again immediately.

> So a function with "wait" in its name doesn't wait sometimes? This needs to be documented in the Doxygen comments. There needs to be enough documentation that a new user of the library can understand why they might need to implement this operation, and how to implement it properly.

Yes.